pkg/query-service/app/integrations/messagingQueues/kafka/sql.go

```diff
@@ -0,0 +1,171 @@
+package kafka
+
+import (
+	"fmt"
+)
+
+func generateConsumerSQL(start, end int64, topic, partition, consumerGroup, queueType string) string {
+	query := fmt.Sprintf(`
+WITH 
+-- Sub query for p99 calculation
+p99_query AS (
+    SELECT
+        serviceName,
+        quantile(0.99)(durationNano) / 1000000 as p99
+    FROM signoz_traces.signoz_index_v2
+    WHERE
+        timestamp >= '%d'
+      AND timestamp <= '%d'
+      AND kind = 5
+      AND msgSystem = '%s' 
+      AND stringTagMap['messaging.destination.name'] = '%s'
+      AND stringTagMap['messaging.destination.partition.id'] = '%s'
+	  AND stringTagMap['messaging.kafka.consumer.group'] = '%s'
+    GROUP BY serviceName
+),
+
+-- Sub query for RPS calculation
+rps_query AS (
+    SELECT
+        serviceName,
+        count(*) / ((%d - %d) / 1000000000) AS rps  -- Convert nanoseconds to seconds
+    FROM signoz_traces.signoz_index_v2
+    WHERE
+        timestamp >= '%d'
+        AND timestamp <= '%d'
+        AND kind = 5
+		AND msgSystem = '%s' 
+        AND stringTagMap['messaging.destination.name'] = '%s'
+        AND stringTagMap['messaging.destination.partition.id'] = '%s'
+		AND stringTagMap['messaging.kafka.consumer.group'] = '%s'
+    GROUP BY serviceName
+),
+
+-- Sub query for error rate calculation
+error_rate_query AS (
+    SELECT
+        serviceName,
+        count(*) / ((%d - %d) / 1000000000) AS error_rate  -- Convert nanoseconds to seconds
+    FROM signoz_traces.signoz_index_v2
+    WHERE
+        timestamp >= '%d'
+        AND timestamp <= '%d'
+        AND statusCode = 2
+        AND kind = 5
+        AND msgSystem = '%s' 
+        AND stringTagMap['messaging.destination.name'] = '%s'
+        AND stringTagMap['messaging.destination.partition.id'] = '%s'
```

Comment:
Some of these subqueries should be combined to reduce the amount of data we read otherwise each query will process the same amount of bytes multiple times. For example, `error_rate_query` and `avg_msg_size_query` can be combined. I would suggest we reduce the bytes we scan as much as possible.

1.
```
WITH
    avg_msg_size_query AS
    (
        SELECT
            serviceName,
            avg(numberTagMap['messaging.message.body.size']) AS avg_msg_size
        FROM signoz_traces.signoz_index_v2
        WHERE ((stringTagMap['messaging.destination.name']) != '') AND (timestamp >= (now() - toIntervalHour(1)))
        GROUP BY serviceName
    ),
    p99_query AS
    (
        SELECT
            serviceName,
            quantile(0.99)(durationNano) / 1000000 AS p99
        FROM signoz_traces.signoz_index_v2
        WHERE ((stringTagMap['messaging.destination.name']) != '') AND (timestamp >= (now() - toIntervalHour(1)))
        GROUP BY serviceName
    )
SELECT *
FROM avg_msg_size_query
INNER JOIN p99_query ON avg_msg_size_query.serviceName = p99_query.serviceName

Query id: 898c1643-6d1e-4a7b-829d-f60e7032a94a

Ok.

0 rows in set. Elapsed: 0.141 sec. Processed 360.84 thousand rows, 138.43 MB (2.56 million rows/s., 982.57 MB/s.)
Peak memory usage: 19.95 MiB.
```

2.

```
WITH avg_msg_size_query AS
    (
        SELECT
            serviceName,
            avg(numberTagMap['messaging.message.body.size']) AS avg_msg_size,
            quantile(0.99)(durationNano) / 1000000 AS p99
        FROM signoz_traces.signoz_index_v2
        WHERE ((stringTagMap['messaging.destination.name']) != '') AND (timestamp >= (now() - toIntervalHour(1)))
        GROUP BY serviceName
    )
SELECT *
FROM avg_msg_size_query

Query id: 7110f972-5008-4abd-abd1-624399c90754

Ok.

0 rows in set. Elapsed: 0.094 sec. Processed 181.00 thousand rows, 72.20 MB (1.92 million rows/s., 767.35 MB/s.)
Peak memory usage: 17.15 MiB.
```

See the `138.43 MB` and `72.20 MB` difference.